### PR TITLE
Thoth s2i advise

### LINKED ISSUE #8: How to use thamos advise cleverness in your s2i Build process

### PR DESCRIPTION
**Pull request by saisankargochhayat**
Closes #8 
This tutorial demos using thoth's advise engine during the s2i build process. 
Link to the tutorial - [link](https://www.katacoda.com/saisankargochhayat/scenarios/hello-world)

### CODE REVIEW

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
+Now that you are familiar with the configurations our s2i image offers, let's try deploying our sample app using Thoth's S2I build process. 
+
+Assuming you have followed the steps from before are logged in as an admin into `myproject` in the openshift cluster lets deploy - 
+
+``oc process -f https://raw.githubusercontent.com/<your-github-username>/s2i-example-tensorflow/log-thoth/openshift.yaml | oc apply -f -``
```
Comment (4n4nd) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: Can you please make this executable by adding {{execute}} at the end like in step 1

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+If you haven't forked the repo and made any changes, you can try deploy our version using - 
+
+``oc process -f https://raw.githubusercontent.com/thoth-station/s2i-example-tensorflow/log-thoth/openshift.yaml | oc apply -f -``{{copy}}
```
Comment (4n4nd) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: {{execute}} instead of {{copy}} might be better

File: ai-machine-learning/thoth-s2i-advise/app.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+Before we explore more about the build process, go ahead fork the sample repo,
+to explore more about the s2i process. 
+
+Here is the example repo we are going to try today - 
+
+``https://github.com/thoth-station/s2i-example-tensorflow``
```
Comment (4n4nd) [inline ai-machine-learning/thoth-s2i-advise/app.md]: here could you make the links clickable?

File: ai-machine-learning/thoth-s2i-advise/app.md [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+
+Now that you have forked it, let's check out the `log-thoth` branch.
+The upstream link to the same is - `https://github.com/thoth-station/s2i-example-tensorflow/tree/log-thoth`
```
Comment (4n4nd) [inline ai-machine-learning/thoth-s2i-advise/app.md]: one more link

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+Make sure you have selected `myproject` on the project selector. 
+If you go to Builds in the Openshift UI in the other tab, under `Builds`, you would see `s2i-example-log-` and under logs you could inspect the build process. 
+You would see `thamos advise` being run your stack and if there is a suggestion. 
+Incase the analysis fails, we resort to the existing Piplock for to prevent the build from failing. 
+
+Your app should be running under `Workloads -> Pods`.
+
+If you want to pull down the remove app you deployed - 
+
+``oc delete all --selector 'app=s2i-example-log'``{{copy}}
```
Comment (4n4nd) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: {{execute}}

File: ai-machine-learning/thoth-s2i-advise/intro.md [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+
+So when you use, thoth s2i build process instead of the normal s2i build process, 
+thoth produces recommendations targeting your specific hardware configuration you 
+use to run your application inside the cluster (e.g. specific GPU available in 
+the cluster).
+
+You can find a list of base images which you can use with Thoth in [s2i-thoth repository](https://github.com/thoth-station/s2i-thoth) 
+with detailed instructions on how to use Thoth in the OpenShift’s s2i process. 
+The container images are hosted at - 
+`quay.io/organization/thoth-station` with the
```
Comment (4n4nd) [inline ai-machine-learning/thoth-s2i-advise/intro.md]: link

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+Now that you have scheduled it in the katacoda terminal on the right, you should see this - 
+```
+buildconfig.build.openshift.io/s2i-example-log created
+deploymentconfig.apps.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-thoth-ubi8-py36 created
+```
+
+### Let's go to Openshift UI and checkout our build process -
```
Comment (4n4nd) [inline ai-machine-learning/thoth-s2i-advise/deploy.md:20]: Do you think adding CLI instructions would be helpful?
For example `oc logs bc/s2i-example-log -f` should show the live log stream in the terminal

**Review by 4n4nd — CHANGES_REQUESTED** [review]

File: ai-machine-learning/thoth-s2i-advise/intro.md [hunk 8]
```diff
@@ -0,0 +1,33 @@
+In this tutorial, we are going to build a simple python app, using Thoth s2i.
+Thus getting intelligent recommendation on the software stack during the build process.
+
+## Why to use thoth s2i build process?
+
+For those who are new to this [s2i](https://docs.openshift.com/container-platform/3.11/using_images/s2i_images/python.html) refers to the Source to image process which
+bundles your source code to a image that can be run on Openshift. 
+
+So when you use, thoth s2i build process instead of the normal s2i build process,
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/intro.md]: Let's name it Thoth with capital T. :pray: 

Here and other places as well.

File: ai-machine-learning/thoth-s2i-advise/intro.md [hunk 9]
```diff
@@ -0,0 +1,33 @@
+In this tutorial, we are going to build a simple python app, using Thoth s2i.
+Thus getting intelligent recommendation on the software stack during the build process.
+
+## Why to use thoth s2i build process?
+
+For those who are new to this [s2i](https://docs.openshift.com/container-platform/3.11/using_images/s2i_images/python.html) refers to the Source to image process which
+bundles your source code to a image that can be run on Openshift.
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/intro.md]: Open**S**hift

File: ai-machine-learning/thoth-s2i-advise/app.md [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+to explore more about the s2i process. 
+
+Here is the example repo we are going to try today - 
+
+``https://github.com/thoth-station/s2i-example``
+
+Now that you have forked it, let's check out the `log-thoth` branch.
+The upstream link to the same is - `https://github.com/thoth-station/s2i-example/tree/log-thoth`
+
+If you go to app.py, it's a simple python app that prints `Hello thoth` every 10 seconds.
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/app.md]: **P**ython

File: ai-machine-learning/thoth-s2i-advise/deploy.md (hunk 1 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: So the users are supposed to replace their username here so I think it's better off as a copy command

File: ai-machine-learning/thoth-s2i-advise/app.md [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+    spec:
+      tags:
+        - name: "latest"
+          from:
+            kind: "DockerImage"
+            name: "quay.io/thoth-station/s2i-thoth-ubi8-py36"
+          referencePolicy:
+            type: "Source"
+```
+And we use this under `build config` under line 26, stating our source repo to be -
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/app.md]: BuildConfig? and maybe no backticks?

File: ai-machine-learning/thoth-s2i-advise/parameter.md [hunk 12]
```diff
@@ -0,0 +1,27 @@
...
+`THOTH_ASSEMBLE_DEBUG` - run s2i’s assemble script in verbose mode
+
+`THOTH_DRY_RUN` - submit stack to Thoth’s recommendation engine but do not use the recommended Pipfile.lock file, use the Pipfile.lock file present in the repo instead
+
+`THOTH_FROM_MASTER` - Use Thamos from git instead of a PyPI release - handy if the released Thamos has a bug which was fixed in the master branch
+
+`THOTH_HOST` - Thoth’s host to reach out to for recommendations (defaults to prod deployment at khemenu.thoth-station.ninja)
+
+`THOTH_ERROR_FALLBACK` - fallback to the Pipfile.lock present in the repository if the submitted Thoth analysis fails
+
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/parameter.md:18]: `THOTH_PROVENANCE_CHECK` (should be always 0 in this scenario?)

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 13]
```diff
@@ -0,0 +1,31 @@
+Now that you are familiar with the configurations our s2i image offers, let's try deploying our sample app using Thoth's S2I build process. 
+
+Assuming you have followed the steps from before are logged in as an admin into `myproject` in the openshift cluster lets deploy - 
+
+``oc process -f https://raw.githubusercontent.com/<your-github-username>/s2i-example/log-thoth/openshift.yaml | oc apply -f -``
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: Maybe remove `<your-github-username>` and keep it pointing to thoth-station? Users could use the link/example directly without a fork.

EDIT: oh, I see the note bellow - I would say its fine to keep pointing/using just thoth-station.

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+buildconfig.build.openshift.io/s2i-example-log created
+deploymentconfig.apps.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-thoth-ubi8-py36 created
+```
+
+### Let's go to Openshift UI and checkout our build process - 
+
+Make sure you have selected `myproject` on the project selector. 
+If you go to Builds in the Openshift UI in the other tab, under `Builds`, you would see `s2i-example-log-` and under logs you could inspect the build process.
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: `s2i-example-log-` the dash is not needed.

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 15]
```diff
@@ -0,0 +1,31 @@
...
+imagestream.image.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-thoth-ubi8-py36 created
+```
+
+### Let's go to Openshift UI and checkout our build process - 
+
+Make sure you have selected `myproject` on the project selector. 
+If you go to Builds in the Openshift UI in the other tab, under `Builds`, you would see `s2i-example-log-` and under logs you could inspect the build process. 
+You would see `thamos advise` being run your stack and if there is a suggestion. 
+Incase the analysis fails, we resort to the existing Piplock for to prevent the build from failing.
```
Comment (fridex) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: s/Piplock/Pipfile.lock/

**Review by fridex — COMMENTED** [review]
Nice! Few minor comments.

File: ai-machine-learning/thoth-s2i-advise/parameter.md (hunk 12 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/parameter.md:18]: I agree! Will put it in the branch also.

File: ai-machine-learning/thoth-s2i-advise/deploy.md (hunk 13 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: Umm, so should I remove that?
I thought users might want to experiment with their forks, but might be too extensive from a tutorial POV

File: ai-machine-learning/thoth-s2i-advise/parameter.md (hunk 12 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/parameter.md:18]: Should we mention it here?

File: ai-machine-learning/thoth-s2i-advise/deploy.md (hunk 7 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md:20]: Thanks added it :100:

File: ai-machine-learning/thoth-s2i-advise/app.md (hunk 3 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/app.md]: Done thanks!

File: ai-machine-learning/thoth-s2i-advise/deploy.md (hunk 13 above)
Comment (fridex) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: IMHO the tutorial should walk a user through the repo and what's present in there. Once they go through it, they are skilled enough to work on the fork.

Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: Makes sense, I will remove it :)

Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: Or how about this - 
If you haven't forked the repo and made any changes, you can try deploy our version using - 

```
``oc process -f https://raw.githubusercontent.com/thoth-station/s2i-example/log-thoth/openshift.yaml | oc apply -f -``{{execute}}

Else if you wanna try with the changes you made on your fork, try this - 
``oc process -f https://raw.githubusercontent.com/<your-github-username>/s2i-example/log-thoth/openshift.yaml | oc apply -f -``{{copy}}

```

File: ai-machine-learning/thoth-s2i-advise/deploy.md (hunk 1 above)
Comment (4n4nd) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: oh yeah makes sense

File: ai-machine-learning/thoth-s2i-advise/deploy.md (hunk 13 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: @fridex I removed instructions to fork, please check.

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 16]
```diff
@@ -0,0 +1,46 @@
+Now that you are familiar with the configurations our s2i image offers, let's try deploying our sample app using Thoth's S2I build process. 
+
+Assuming you have followed the steps from before are logged in as an admin into `myproject` in the openshift cluster lets deploy -
```
Comment (hemajv) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: minor fix: "Assuming you have followed the steps from before `and` are logged in as an admin..."

Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: Thanks for pointing that :) Added in - [b2208fe](https://github.com/AICoE/learn-katacoda/pull/21/commits/b2208fe6a73f80d0bc19ff4e63b205fe5c82f1d4)

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 17]
```diff
@@ -0,0 +1,46 @@
+Now that you are familiar with the configurations our s2i image offers, let's try deploying our sample app using Thoth's S2I build process. 
+
+Assuming you have followed the steps from before and are logged in as an admin into `myproject` in the openshift cluster lets deploy - 
+
+If you haven't forked the repo and made any changes, you can try deploy our version using -
```
Comment (hemajv) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: change to: "you can try **_deploying_** our version using..."?

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+deploymentconfig.apps.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-thoth-ubi8-py36 created
+```
+
+### Let's go to Openshift UI and checkout our build process - 
+
+Make sure you have selected `myproject` on the project selector. 
+If you go to Builds in the Openshift UI in the other tab, under `Builds`, you would see `s2i-example-log` and under logs you could inspect the build process. 
+You would see `thamos advise` being run your stack and if there is a suggestion.
```
Comment (hemajv) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: change to: "being run **on** your stack"?

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+imagestream.image.openshift.io/s2i-example-log created
+imagestream.image.openshift.io/s2i-thoth-ubi8-py36 created
+```
+
+### Let's go to Openshift UI and checkout our build process - 
+
+Make sure you have selected `myproject` on the project selector. 
+If you go to Builds in the Openshift UI in the other tab, under `Builds`, you would see `s2i-example-log` and under logs you could inspect the build process. 
+You would see `thamos advise` being run your stack and if there is a suggestion. 
+Incase the analysis fails, we resort to the existing Pipfile.lock for to prevent the build from failing.
```
Comment (hemajv) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: remove "**for**"?

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+
+Now lets check the logs - 
+
+``oc logs bc/s2i-example-log -f``{{execute}}
+
+You should keep a eye for these things in the log - 
+ - Thoth's configuration file after hardware and software discovery (that's the .thoth.yaml being expanded from the template.)
+ - Asking Thoth for advise... (That is where thamos interacts with Thoth API)
+
+Now if check the UI, your app should be running under `Workloads -> Pods`.
```
Comment (hemajv) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: add: "now if **you** check the UI..."

File: ai-machine-learning/thoth-s2i-advise/deploy.md [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+
+Once the application is deployed, you can check the logs from the deployed app using - 
+
+``oc logs -f dc/s2i-example-log``{{execute}}
+
+After Thoth finishes processing your stack you should see something similar to this in your report - 
+![thoth advise pass](https://raw.githubusercontent.com/saisankargochhayat/katacoda-scenarios/master/hello-world/assets/thamos_advise_pass.png)
+
+If you want to pull down the remove app you deployed -
```
Comment (hemajv) [inline ai-machine-learning/thoth-s2i-advise/deploy.md]: did you mean to say: "If you want to remove the app you deployed?"

File: ai-machine-learning/thoth-s2i-advise/finish.md [hunk 22]
```diff
@@ -0,0 +1,7 @@
+Thank you for trying to build usinng Thoth s2i build process.
```
Comment (hemajv) [inline ai-machine-learning/thoth-s2i-advise/finish.md]: maybe rephrase to: "Thank you for trying to create an application using the Thoth's s2i build process."?

**Review by hemajv — COMMENTED** [review]
Apart from these minor fixes, lgtm! :ship: 
Nice scenario! :+1:

File: ai-machine-learning/thoth-s2i-advise/finish.md (hunk 22 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/finish.md]: Thanks I have changed it :)

File: ai-machine-learning/thoth-s2i-advise/deploy.md (hunk 21 above)
Comment (saisankargochhayat) [reply ai-machine-learning/thoth-s2i-advise/deploy.md]: Thanks I have changed it :)

**Comment by 4n4nd** [comment]
LGTM :+1: 
Just a heads up, in the final step when we try to see the dc logs, 
If you try to see the logs while the application is still being deployed, it will show you the deployment scaling logs.
I ran the command twice to see the application logs
Screenshot from 2020-04-07 00-50-40

**Comment by saisankargochhayat** [comment]
> final step when we try to see the dc logs,
> If you try to see the logs while the application is still being deployed, it will show you the deployment scaling logs.

I was hoping they would run bc log and when it ends, they would run dc log.

**Comment by 4n4nd** [comment]
> > final step when we try to see the dc logs,
> > If you try to see the logs while the application is still being deployed, it will show you the deployment scaling logs.
> 
> I was hoping they would run bc log and when it ends, they would run dc log.

That's what I did, but instead of the pod logs it showed me the scaling logs because the pod was still being deployed.

**Comment by saisankargochhayat** [comment]
> > > final step when we try to see the dc logs,
> > > If you try to see the logs while the application is still being deployed, it will show you the deployment scaling logs.
> > 
> > 
> > I was hoping they would run bc log and when it ends, they would run dc log.
> 
> That's what I did, but instead of the pod logs it showed me the scaling logs because the pod was still being deployed.

image
So like after the build is completed, it shows `Push successful`
Then you do dc logs.

**Comment by 4n4nd** [comment]
> > > > final step when we try to see the dc logs,
> > > > If you try to see the logs while the application is still being deployed, it will show you the deployment scaling logs.
> > > 
> > > 
> > > I was hoping they would run bc log and when it ends, they would run dc log.
> > 
> > 
> > That's what I did, but instead of the pod logs it showed me the scaling logs because the pod was still being deployed.
> 
> image
> So like after the build is completed, it shows `Push successful`
> Then you do dc logs.

After the image is pushed, the deployment takes some time to be ready (spinning up a pod takes some time), while the pod is being created you will see the scaling log instead of the pod log. But I guess that's not a big deal :+1:

**Review by 4n4nd — APPROVED** [review]
:+1: LGTM
